File: simplified-profiles-api/src/main/java/org/nypl/simplified/profiles/api/ProfilePreferences.java
package org.nypl.simplified.profiles.api;

import com.google.auto.value.AutoValue;
import com.io7m.jfunctional.Option;
import com.io7m.jfunctional.OptionType;

import org.joda.time.LocalDate;
import org.nypl.simplified.reader.api.ReaderPreferences;

/**
 * A set of preferences for a profile.
 */

@AutoValue
public abstract class ProfilePreferences {

  ProfilePreferences() {

  }

  /**
   * @return The gender of the reader (if one has been explicitly specified)
   */

  public abstract OptionType<String> gender();

  /**
   * @return The date of birth of the reader (if one has been explicitly specified)
   */

  public abstract OptionType<ProfileDateOfBirth> dateOfBirth();

  /**
   * @return The current value as a mutable builder
   */

  public abstract Builder toBuilder();

  /**
   * @return The reader-specific preferences
   */

  public abstract ReaderPreferences readerPreferences();

  /**
   * A mutable builder for the type.
   */

  @AutoValue.Builder
  public abstract static class Builder {

    Builder() {

    }

    /**
     * @param prefs The reader preferences
     * @return The current builder
     * @see #readerPreferences()
     */

    public abstract Builder setReaderPreferences(
      ReaderPreferences prefs);

    /**
     * @param gender The gender
     * @return The current builder
     * @see #gender()
     */

    public abstract Builder setGender(
      OptionType<String> gender);

    /**
     * @param gender The gender
     * @return The current builder
     * @see #gender()
     */

    public final Builder setGender(final String gender) {
      return setGender(Option.some(gender));
    }

    /**
     * @param date The date
     * @return The current builder
     * @see #dateOfBirth()
     */

    public abstract Builder setDateOfBirth(
      OptionType<ProfileDateOfBirth> date);

    /**
     * @param date The date
     * @return The current builder
     * @see #dateOfBirth()
     */

    public final Builder setDateOfBirth(final ProfileDateOfBirth date) {
      return setDateOfBirth(Option.some(date));
    }

    /**
     * @return A profile description based on the given parameters
     */

    public abstract ProfilePreferences build();
  }

  /**
   * @return A new builder
   */

  public static ProfilePreferences.Builder builder() {
    return new AutoValue_ProfilePreferences.Builder()
      .setReaderPreferences(ReaderPreferences.builder().build())
      .setGender(Option.none())
      .setDateOfBirth(Option.none());
  }
}

File: simplified-profiles-api/src/main/java/org/nypl/simplified/profiles/api/ProfilePreferences.kt
package org.nypl.simplified.profiles.api

import org.nypl.simplified.accounts.api.AccountID
import org.nypl.simplified.reader.api.ReaderPreferences

/**
 * The preferences for a profile.
 *
 * A _preference_ is distinct from an _attribute_ in that the application is permitted to
 * behave differently based upon the value of a _preference_. Preferences are typed.
 *
 * @see ProfileAttributes
 */

data class ProfilePreferences(

  /** @return The date of birth of the reader (if one has been explicitly specified) */

  val dateOfBirth: ProfileDateOfBirth?,

  /** @return `true` if non-production libraries should be displayed */

  val showTestingLibraries: Boolean,

  /**
   * @return `true` if the user has seen the library selection screen and opted out of
   * selecting a library
   */

  val hasSeenLibrarySelectionScreen: Boolean,

  /** @return The reader-specific preferences */

  val readerPreferences: ReaderPreferences,

  /** The most recently used account provider. */

  val mostRecentAccount: AccountID,

  /** @return `true` if the debug settings should be visible. */

  val showDebugSettings: Boolean = false
)
